torch/csrc/lazy/core/ir.cpp

```diff
@@ -52,28 +79,115 @@ Node::Node(OpKind op, size_t num_outputs, hash_t node_hash, std::function<hash_t
       node_hash_(node_hash),
       dag_hash_without_sizes_(dag_hash_fn(false)),
       dag_hash_with_sizes_(dag_hash_fn(true)),
-      metadata_(GetMetaDataIfDebugging()) {}
+      metadata_(GetMetaDataIfDebugging()),
+      python_stacktrace_(GetFirstUserFrameInPythonIfEnabled()) {}
 
 Node::Node(OpKind op, size_t num_outputs, std::function<hash_t(bool)> node_hash_fn)
     : op_(op),
       num_outputs_(num_outputs),
       node_hash_(node_hash_fn(!enableDynamicShape())),
       dag_hash_without_sizes_(node_hash_fn(false)),
       dag_hash_with_sizes_(node_hash_fn(true)),
-      metadata_(GetMetaDataIfDebugging()) {}
+      metadata_(GetMetaDataIfDebugging()),
+      python_stacktrace_(GetFirstUserFrameInPythonIfEnabled()) {}
+
+
+Node::Node(OpKind op, OpList operands, std::vector<Shape>&& shapes,
+           size_t num_outputs, hash_t hash_seed)
+    : Node(op, num_outputs,
+           // TODO(WHC) this is inefficient (having to compute node_hash twice
+           // since I can't call hash() yet) so probably move dag_hash
+           // initialization to a separate function?
+           /* node_hash */ HashCombine(op.hash(), hash_seed),
+           /* dag_hash */
+           [&](bool bakeInSizes) { return OperandHashes(operands, HashCombine(op.hash(), hash_seed), bakeInSizes); }
+           ) {
+  shapes_.insert(
+    shapes_.end(),
+    std::make_move_iterator(shapes.begin()), 
+    std::make_move_iterator(shapes.end()));
+  for (auto& operand : operands) {
+    // Ideally, optional operands should be filtered by the leaf node classes,
+    // but it's just much easier to do it here.
+    // TODO(alanwaketan): Find a way to move the below logic to the leaf node
+    // classes.
+    if (!operand) {
+      continue;
+    }
+
+    AddOperand(operand.node, operand.index);
+  }
+}
+
+Node::Node(OpKind op, OpList operands,
+           const std::function<Shape()>& shape_fn,
+           size_t num_outputs, hash_t hash_seed)
+    : Node(op, operands, std::vector<Shape>{}, num_outputs, hash_seed) {
+  shapes_.push_back(GetOpShape(shape_fn));
+}
+
+Node::Node(OpKind op, OpList operands, size_t num_outputs,
+           hash_t hash_seed)
+    : Node(op, operands, std::vector<Shape>{}, num_outputs, hash_seed) {}
+
+Node::Node(OpKind op, Shape shape, size_t num_outputs, hash_t hash_seed)
+    : Node(op, num_outputs, [&](bool bakeInSizes) -> hash_t { return GetOpHash(op, shape, hash_seed, bakeInSizes); })
+{
+  shapes_.push_back(std::move(shape));
+}
 
 Node::~Node() = default;
 
+void Node::SetShapeDeferred(
+    const std::function<Shape()>& shape_fn) {
+  shapes_.push_back(GetOpShape(shape_fn));
+}
+
+const Shape& Node::shape(size_t output_index) const {
+  return shapes_.at(output_index);
+}
+
+using ShapeCache = Cache<hash_t, Shape, HashReducer>;
+
+ShapeCache* GetShapeCache() {
+  static ShapeCache* cache = new ShapeCache(FLAGS_torch_lazy_tensors_shape_cache_size);
+  return cache;
+}
+
+Shape Node::GetOpShape(
```

Comment:
@shunting314 can you comment on how you would prefer to handle shape cache and GetOpShape API given your other refactor plans?

Reply:
Since the PR move shape cache to backend agnostic folder, I'm thinking if we can add a pure virtual method to Node called hash_for_shape_cache() ? 

XLA can simply make hash_for_shape_cache / hash identical. But TS backend will have different implementation for them. Does that make sense? cc @wconstab  @JackCaoG